### PR TITLE
New Variant FK407M1

### PR DESCRIPTION
**Pull request by BennehBoy**
Chinese STM32F407VET6 Minimum Development board with USB-C connector and SDIO.

Untested.

### CODE REVIEW

**Comment by BennehBoy** [comment]
20-40 days for my boards to ship, oh well 🤣

**Comment by fpistm** [comment]
Do you have a link on this board and the schematics?

**Comment by BennehBoy** [comment]
Blink tested - found issue with PC13 definition (pin is not broken out)

HID bootloader tested

USART, SDIO, SPI to test...

**Comment by BennehBoy** [comment]
USART working (1 & 3)
CDC working
SPI1 working
SDIO working

**Comment by BennehBoy** [comment]
I can't get i2c scanner to report anything, but not just with this board, same problem with 407VE.

The device I'm testing against is found by scanner against rogers core on a maple mini - same wiring.

I'll try and test on bluepill & blackpill shortly.

I'm not forgetting something daft like pin initialisation?  Wire lib should take care of that right?

**Comment by BennehBoy** [comment]
```
Scanning...
Unknown error at address 0x02
Unknown error at address 0x03
Unknown error at address 0x04
Unknown error at address 0x05
Unknown error at address 0x06
Unknown error at address 0x07
Unknown error at address 0x08
Unknown error at address 0x09
Unknown error at address 0x0A
Unknown error at address 0x0B
Unknown error at address 0x0C
Unknown error at address 0x0D
Unknown error at address 0x0E
Unknown error at address 0x0F
Unknown error at address 0x10
Unknown error at address 0x11
Unknown error at address 0x12
Unknown error at address 0x13
Unknown error at address 0x14
Unknown error at address 0x15
Unknown error at address 0x16
Unknown error at address 0x17
Unknown error at address 0x18
Unknown error at address 0x19
Unknown error at address 0x1A
Unknown error at address 0x1B
Unknown error at address 0x1C
Unknown error at address 0x1D
Unknown error at address 0x1E
Unknown error at address 0x1F
Unknown error at address 0x20
Unknown error at address 0x21
Unknown error at address 0x22
Unknown error at address 0x23
Unknown error at address 0x24
Unknown error at address 0x25
Unknown error at address 0x26
Unknown error at address 0x27
Unknown error at address 0x28
Unknown error at address 0x29
Unknown error at address 0x2A
Unknown error at address 0x2B
Unknown error at address 0x2C
Unknown error at address 0x2D
Unknown error at address 0x2E
Unknown error at address 0x2F
Unknown error at address 0x30
Unknown error at address 0x31
Unknown error at address 0x32
Unknown error at address 0x33
Unknown error at address 0x34
Unknown error at address 0x35
Unknown error at address 0x36
Unknown error at address 0x37
Unknown error at address 0x38
Unknown error at address 0x39
Unknown error at address 0x3A
Unknown error at address 0x3B
Unknown error at address 0x3C
Unknown error at address 0x3D
Unknown error at address 0x3E
Unknown error at address 0x3F
Unknown error at address 0x40
Unknown error at address 0x41
Unknown error at address 0x42
Unknown error at address 0x43
Unknown error at address 0x44
Unknown error at address 0x45
Unknown error at address 0x46
Unknown error at address 0x47
Unknown error at address 0x48
Unknown error at address 0x49
Unknown error at address 0x4A
Unknown error at address 0x4B
Unknown error at address 0x4C
Unknown error at address 0x4D
Unknown error at address 0x4E
Unknown error at address 0x4F
Unknown error at address 0x50
Unknown error at address 0x51
Unknown error at address 0x52
Unknown error at address 0x53
Unknown error at address 0x54
Unknown error at address 0x55
Unknown error at address 0x56
Unknown error at address 0x57
Unknown error at address 0x58
Unknown error at address 0x59
Unknown error at address 0x5A
Unknown error at address 0x5B
Unknown error at address 0x5C
Unknown error at address 0x5D
Unknown error at address 0x5E
Unknown error at address 0x5F
Unknown error at address 0x60
Unknown error at address 0x61
Unknown error at address 0x62
Unknown error at address 0x63
Unknown error at address 0x64
Unknown error at address 0x65
Unknown error at address 0x66
Unknown error at address 0x67
Unknown error at address 0x68
Unknown error at address 0x69
Unknown error at address 0x6A
Unknown error at address 0x6B
Unknown error at address 0x6C
Unknown error at address 0x6D
Unknown error at address 0x6E
Unknown error at address 0x6F
Unknown error at address 0x70
Unknown error at address 0x71
Unknown error at address 0x72
Unknown error at address 0x73
Unknown error at address 0x74
Unknown error at address 0x75
Unknown error at address 0x76
Unknown error at address 0x78
Unknown error at address 0x79
Unknown error at address 0x7A
Unknown error at address 0x7B
Unknown error at address 0x7C
Unknown error at address 0x7D
Unknown error at address 0x7E
No I2C devices found

```

**Comment by BennehBoy** [comment]
Working on Bluepill, so perhaps something common to the GenF4 variants - I cloned the base 407VE to define this?

**Comment by BennehBoy** [comment]
Hmm...

If I force SDA to PB7 and SCL to PB6:

```
  Wire.setSDA(PB7);
  Wire.setSCL(PB6);
```

Then it works....

I thought these were the default pins?

From PeripheralPins.c:

```
#ifdef HAL_I2C_MODULE_ENABLED
const PinMap PinMap_I2C_SDA[] = {
  {PB_7,  I2C1, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C1)},
  {PB_9,  I2C1, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C1)},
  {PB_11, I2C2, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C2)},
  {PC_9,  I2C3, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C3)},
  {NC,    NP,    0}
};
#endif

#ifdef HAL_I2C_MODULE_ENABLED
const PinMap PinMap_I2C_SCL[] = {
  {PA_8,  I2C3, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C3)},
  {PB_6,  I2C1, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C1)},
  {PB_8,  I2C1, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C1)},
  {PB_10, I2C2, STM_PIN_DATA(STM_MODE_AF_OD, GPIO_NOPULL, GPIO_AF4_I2C2)},
  {NC,    NP,    0}
};
#endif
```

**Comment by uzi18** [comment]
this  shows it is better to explicitly set I/Os than think it is set to some defaults

**Comment by uzi18** [comment]
Just thinking about move peripheralpins.c from variant to core, so less to fix if something change and could be more generic, also variant will be more. simple

**Comment by BennehBoy** [comment]
I don't really see how that would work, it's specific to each board so if it was in the core it would need so many switches as to make it unintelligible.

**Comment by uzi18** [comment]
just check stm32generic these variants are realy simple

**Comment by BennehBoy** [comment]
Perhaps, but that's not the way this core is structured - I suspect DianelF had to do a lot of leg work to make the internals work on STM32GENERIC

File: variants/FK407M1/variant.h [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+// Below SPI and I2C definitions already done in the core
+// Could be redefined here if differs from the default one
+// SPI Definitions
+
+#define PIN_SPI_MOSI            PB5 // NRF24 connector & W25Q16 (on board flash)
+#define PIN_SPI_MISO            PB4 // NRF24 connector & W25Q16 (on board flash)
+#define PIN_SPI_SCK             PB3 // NRF24 connector & W25Q16 (on board flash)
+
+// I2C Definitions
+#define PIN_WIRE_SDA            PB9
```
Comment (fpistm) [inline variants/FK407M1/variant.h]: You set the default pins for Wire instance to be PB9 and PB8. If you think it should be PB7/PB6 then update it here.

Comment (BennehBoy) [reply variants/FK407M1/variant.h]: Ah lol OK - sleep deprived, copy and paste from Black407XX

Not sure why that variant chose to go with the alternate I2C1 pins.

**Comment by BennehBoy** [comment]
On that note, analog pins are not consecutive in this variant - they are broken out at seemingly random so PeripheralPins.c would be a bit confusing for anyone looking at it (I think).

File: variants/FK407M1/variant.h (hunk 1 above)
Comment (fpistm) [reply variants/FK407M1/variant.h]: difficult to tell maybe for PCB reasons...

Comment (BennehBoy) [reply variants/FK407M1/variant.h]: I've eyeballed the board and there's no reason why it should not be on the default pins.   Odd one.

**Comment by zoomx** [comment]
@fpistm,
the Arduino compatibility is used only on boards that has these pin name written on PCB. All other boards don't have these names so I believe it's better to use PXn. IT is the same solution used in ESP8266 and TI (Texas Instruments) core.

**Comment by BennehBoy** [comment]
> @fpistm,
> the Arduino compatibility is used only on boards that has these pin name written on PCB. All other boards don't have these names so I believe it's better to use PXn. IT is the same solution used in ESP8266 and TI (Texas Instruments) core.

I agree, but it IS nice to have the integer value of say A0 be the same as it's PXn equivalent - perhaps the readme.md should simply state which boards do and do not have Arduino pin equivalency?

File: variants/FK407M1/variant.h (hunk 1 above)
Comment (fpistm) [reply variants/FK407M1/variant.h]: Just checking layout and schematics, I also see no reason. So, up to you to choose the default I2C pins used by Wire instance.

**Comment by fpistm** [comment]
Thanks  @BennehBoy 
I've rebased it manually and did some clean that's why I did not merge this one.

**Pull request closed without merging**